### ISSUE TITLE
Questions about new indent rule behavior in regards to comments

### ISSUE THREAD

**Issue by kaicataldo** [eslint/eslint#8704]
**Tell us about your environment**

* **ESLint Version:** Master
* **Node Version:** 6.x
* **npm Version:** 4.x

**What parser (default, Babel-ESLint, etc.) are you using?**
Espree

**Please show your full configuration:**

```json
{
  "rules": {
    "indent": ["error", 2]
  }
}

```
**What did you do? Please include the actual source code causing the issue.**
Noticed an inconsistency with how comments are handled by the new indent rule

**What did you expect to happen?**
For this to pass:

```js
function foo() {
  if (true) {
    // some comment
  }
}
```

And for these to not:
```js
function foo() {
  if (true) {
// some comment
  }
}

function foo() {
  if (true) {
  // some comment
  }
}
```

**What actually happened? Please include the actual, raw output from ESLint.**
Both the following code snippets do not throw an error with the above configuration:

```js
function foo() {
  if (true) {
    // some comment
  }
}

function foo() {
  if (true) {
  // some comment
  }
}
```
However, the following example does:
```js
function foo() {
  if (true) {
// some comment
  }
}
```
```
/Users/cataldo/Code/test/eslint-test/index.js
  3:1  error  Expected indentation of 4 spaces but found 0  indent
```

I think there are two issues at play here:
- I believe the error message is correct, and that it should enforce for 4 spaces with the above code snippet
- It's confusing that the error message expects 4 spaces but is satisfied with 2 spaces

**Comment by not-an-aardvark** [comment]
This is working as indended because there are several possible locations where a comment can reasonably appear, so the rule allows comments to be aligned with the preceding or following token. For example, the rule can't tell whether a user meant to do something like this:

```js
if (x) {
    foo();
    // comment about foo()
} else if (bar()) {
    baz();
}
```

Or something like this:

```js
if (x) {
    foo();
// comment about bar()
} else if (bar()) {
    baz();
}
```

**Comment by platinumazure** [comment]
I wouldn't want to change this for 4.0, but maybe could we enhance the rule with an option directing the rule to assume comments should precede or follow tokens, and thus only allow one indentation for a given comment?

**Comment by kaicataldo** [comment]
The problem I'm trying to solve is that I've enabled the new rule in our codebases at work and am able to fix everything but comments. Manually going through and fixing a PR that has `+3,978 −3,991` changes is difficult.

I wonder if we could make the rule smarter by saying that it's safe to assume the comment should be indented if it's at the top of a block?

I have a lot of instances where the following:
```js
function foo() {
  return bar
 .then((baz) => {
  // some comment
   return qux(baz);
 }
}
```
get autofixed to:
```js
function foo() {
  return bar
   .then((baz) => {
  // some comment
     return qux(baz);
   }
}
```
It seems like a safe assumption that the comment should also be indented, but maybe I'm missing something here!

**Comment by kaicataldo** [comment]
Oh, and one final thing - I do feel like the error message is a bug, since it doesn't actually reflect the possible fix options (2 or 4 spaces, in the example provided in the original post). Is that something we can make better?

**Comment by not-an-aardvark** [comment]
> I wouldn't want to change this for 4.0, but maybe could we enhance the rule with an option directing the rule to assume comments should precede or follow tokens, and thus only allow one indentation for a given comment?

Possibly, but I'm unsure whether this option would be ergonomic. In my view, the problem isn't that people have multiple stylistic preferences for this case, it's that the (presumably shared) stylistic preference *depends on what the comment is referring to*. So a project configures the rule to always allow one indentation for a given comment, it will necessarily end up with some comments having the "wrong" indentation (assuming they want the comment to be aligned with the thing that it refers to). Effectively, we would be enforcing a suboptimal indentation just for the sake of consistency, which seems like it could be annoying for users.

> I wonder if we could make the rule smarter by saying that it's safe to assume the comment should be indented if it's at the top of a block?

I'm a bit confused by the example you provided -- it seems like the comment isn't in a valid location either way (it's indented by 1 space ahead of `.then`, but 1 space behind `return`). Is there a typo in the example?

> Oh, and one final thing - I do feel like the error message is a bug, since it doesn't actually reflect the possible fix options (2 or 4 spaces, in the example provided in the original post). Is that something we can make better?

I wouldn't be opposed to this. My one concern is that this is sort of an edge case (most of the time, the "correct" indentation for the comment is the value in the report message), so we should make sure we don't make the message more confusing for the common case where the current message is correct.

**Comment by kaicataldo** [comment]
@not-an-aardvark Sorry, there was a typo (tried to type this up in the comment instead of an environment with fixed width font - should have double checked). I updated the comment with the correct indentation.

As to the error messaging, I think you make a good point. Maybe it's best to leave it as is for now and see if there are any reports/suggestions going forward.

Removed the bug label because everything is working as intended. Thanks for the discussion!

**Comment by yairEO** [comment]
@kaicataldo - why did you close this? I also am looking for a way to ignore comments indentation in our code. I can't possibly fix hundreds of comments in dozens of files..

**Comment by not-an-aardvark** [comment]
@yairEO This is working as intended. If you'd like a version of the `indent` rule that is less strict and ignores comments, you can use the `indent-legacy` rule.

**Comment by yairEO** [comment]
@not-an-aardvark - Thanks! that's awesome! why doesn't it say in the [docs](http://eslint.org/docs/4.0.0/rules/indent-legacy) about the `legacy` ignoring comments?? people are googling that and will never find out that the `legacy` rule does that if it wouldn't be mentioned so in the docs.

**Comment by not-an-aardvark** [comment]
The `indent-legacy` rule is deprecated because it had many bugs that would cause it to ignore certain patterns in code, such as comments. In most cases, ignoring the indentation of something is undesirable, so we generally recommend people use the `indent` rule instead.
